### ISSUE TITLE
VB -> C#: variable is initialized inside loop

### ISSUE THREAD

**Issue by KunzeAndreas** [icsharpcode/CodeConverter#897]
### VB.Net input code
```vbnet
Private Shared Sub DefaultInitialization()
    For i = 1 To 2
        Dim b As Boolean
        If i = 1 Then
            Assert.That(Not b)
        Else
            Assert.That(b)
        End If

       b = True
       Assert.That(b)
   Next
 End Sub
```

### Erroneous output
```C#
private static void DefaultInitialization()
{
    for (int i = 1; i <= 2; i++)
    {
        var b = default(bool);
        if (i == 1)
        {
            Assert.That(!b);
        }
        else
        {
            Assert.That(b);
        }

        b = true;
        Assert.That(b);
    }
}
```

### Expected output
```C#
private static void DefaultInitialization()
{
    var b = default(bool);  // define and initialize outside loop or do not initialize (with compile error CS0165)
    for (int i = 1; i <= 2; i++)
    {
        // bool b;
        if (i == 1)
        {
            Assert.That(!b);
        }
        else
        {
            Assert.That(b);
        }

        b = true;
        Assert.That(b);
    }
}
```

### Details
* Product in use: VS extension 
* Version in use: 9.0.0.0
* Did you see it working in a previous version, which? no

**Comment by GrahamTheCoder** [comment]
VB still manages to surprise me after all these years. If I ever knew about that "feature" of VB (pulling the variable into the outer scope), I've erased it from my mind. I guess it's because VB only allows one loop variable but you might want multiple, and there's no other way to limit the scope to within the loop since blocks don't exist. Whatever the reason, yep it definitely needs fixing in the converter!

**Comment by GrahamTheCoder** [comment]
For reference, it's a compile error to clash with the name in the enclosing block: [sharplab example](https://sharplab.io/#v2:DYLgbgRgNALiBOBXAdlAJiA1AHwJIFsAHAe3hgGcACAZQE9yYBTfAWACh2AFRCYASwDGlAMLAAhuSoANdpTmVuvQTQAWY+IzQ0elALJi+yABQBKWfIsAxUpT6UAvJQCMlACrFKAZnMXflACJ8+JQQlACCVLjITADmjPAOzj5+8qGUmPZOySkixMjkxMCMAHQA6vB8TAAyhoxGEGZsKQByjAAeMMmBwaERlABCxIWMYsjJAKLIWtQ87JNaohLkQA=), which means it's pretty safe to pull uninitialized variable declarations out.

Separately from whatever intiial fix we do, in the case where the first usage in the loop is *definitely* a write usage, we *could* decide to leave the declaration insdie the loop for simplicity, neatness and similarity in appearance/scoping to the source

**Comment by Noah1989** [comment]
Sounds like another job for PerScopeState / IHoistedNode

**Comment by Noah1989** [comment]
> For reference, it's a compile error to clash with the name in the enclosing block: [sharplab example](https://sharplab.io/#v2:DYLgbgRgNALiBOBXAdlAJiA1AHwJIFsAHAe3hgGcACAZQE9yYBTfAWACh2AFRCYASwDGlAMLAAhuSoANdpTmVuvQTQAWY+IzQ0elALJi+yABQBKWfIsAxUpT6UAvJQCMlACrFKAZnMXflACJ8+JQQlACCVLjITADmjPAOzj5+8qGUmPZOySkixMjkxMCMAHQA6vB8TAAyhoxGEGZsKQByjAAeMMmBwaERlABCxIWMYsjJAKLIWtQ87JNaohLkQA=), which means it's pretty safe to pull uninitialized variable declarations out.

It's not that simple. We need some name mangling, because multiple for loops with the same variable name end up actually creating separate variables: [sharplab example](https://sharplab.io/#v2:DYLgbgRgNALiBOBXAdlAJiA1AHwJIFsAHAe3hgGcACAZQE9yYBTfAWACh2AFRCYASwDGlAMLAAhuSoANdpTmVuvQTQAWY+IzQ0elALJi+yABQBKWfPkAxUpT6UAvJQCMlACrFKAZnMXflACJ8+JQQlACCVLjITADmjPA+fvKhlJj2TolJIsTI5MTAjAB0AOrwfEwAMoaMRgBEEPYA3gAMAL61UBBmbFkW1vCUAFYOzm4e3j29foHBQhGUUbHxmVOUQqnpK1PCOXkFJWWV1XUCTk1tHQLdq5QAcowAHjBbvv1DIy7uXi9ZM2vhkWijDiCUmNzk6zSGTB4J2uXyRVK5UYVWQNVqAgATOd2lArj95Pcnlsic8ODCAKLILTUHjsKlaUQSchAA===)

**Comment by Noah1989** [comment]
> I think, it would be OK, when we end with a compile error (e.g., variable not initialized) in the converted C# code, so we have the chance to fix this in VB code (pull variable to outer scope).

Generating the correct code is not a problem, since I already did a very similar thing for VB static variables. The detection of these cases is the difficult part.

**Comment by GrahamTheCoder** [comment]
Fixed in next release by #902
